### ISSUE TITLE
Invalid sorting on line break in artifactId value

### ISSUE THREAD

**Issue by vrastov** [Ekryd/sortpom#99]
Settings:
```xml
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.13.1</version>
        <configuration>
          <keepBlankLines>true</keepBlankLines>
          <createBackupFile>false</createBackupFile>
          <expandEmptyElements>false</expandEmptyElements>
          <sortModules>true</sortModules>
          <sortDependencies>scope,groupId,artifactId</sortDependencies>
          <sortProperties>true</sortProperties>
          <sortPlugins>groupId,artifactId</sortPlugins>
          <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
        </configuration>
      </plugin>
```

Before sorting:

```xml
    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>baz</artifactId>
    </dependency>

    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>
        foo.bar.veryLongArtifactId
      </artifactId>
    </dependency>
```
After first sorting: 
```xml
    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>foo.bar.veryLongArtifactId
      </artifactId>
    </dependency>

    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>baz</artifactId>
    </dependency>
```
On restart:
```xml
    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>baz</artifactId>
    </dependency>

    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>foo.bar.veryLongArtifactId
      </artifactId>
    </dependency>
```

But after formatting:
```xml
    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>
        foo.bar.veryLongArtifactId
      </artifactId>
    </dependency>

    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>baz</artifactId>
    </dependency>
```
Does not change on subsequent launches.

But baz I expect first.

**Comment by Ekryd** [comment]
Hi, sound interesting. I'll have a look

**Comment by Ekryd** [comment]
First question: Should we allow line breaks inside elements at all?

I would expect that
```
 <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>
        foo.bar.veryLongArtifactId
      </artifactId>
    </dependency>
```

becomes 
```
 <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>foo.bar.veryLongArtifactId</artifactId>
    </dependency>
```

Do you use any artifactIds that are over 40 characters?

**Comment by vrastov** [comment]
Whitespace is part of the value. https://tools.ietf.org/html/rfc3470#page-18

And I'm kind of wrong, but the maven works :smile:.

And yes, the length is more than 40 characters, a line of 120 characters does not fit. For uniqueness, artifactId include a groupId.

**Comment by Ekryd** [comment]
Whitespace is part of the xml specification and it is allowed within XML, but Maven does not take whitespace into consideration. So line break before or after an artifactId is trimmed by Maven before processing.

You must have some great requirements on uniqueness if you need to put the full groupId inside every artifactId ;-D
It is quite common to use some part of of the groupId (ie the company or organisation name) in the artifactId such as:
```xml
    <dependency>
      <groupId>com.something.or.another.company</groupId>
      <artifactId>company.system.component.part</artifactId>
    </dependency>
```
But that is beside the point. (and not my concern)

I have started to look, and I think that the default behaviour should be to remove all whitespace inside elements, so something may be strange. Digging is not done yet.

**Comment by vrastov** [comment]
Cut off the same part, yes that's brilliant :smile:  , thanks!

**Comment by Ekryd** [comment]
I suspect that something else in your environment is formatting the pom.xml as well. I used this pom as example:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.github.ekryd.sortpom.its</groupId>
  <artifactId>default-configuration</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>
        junit-jupiter-api
      </artifactId>
      <version>5.7.0</version>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.13.1</version>
        <executions>
          <execution>
            <goals>
              <goal>sort</goal>
            </goals>
            <configuration>
              <keepBlankLines>true</keepBlankLines>
              <createBackupFile>false</createBackupFile>
              <expandEmptyElements>false</expandEmptyElements>
              <sortModules>true</sortModules>
              <sortDependencies>scope,groupId,artifactId</sortDependencies>
              <sortProperties>true</sortProperties>
              <sortPlugins>groupId,artifactId</sortPlugins>
              <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```

and executed `mvn clean verify`. The result sorts the dependencies in the right order.

```xml
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.0</version>
    </dependency>
  </dependencies>
```

All the line breaks inside the elements are removed. Do you use more formatting plugins in Maven? Do you have the same behaviour from the command prompt?

**Comment by Ekryd** [comment]
Sortpom should not care about long lines. It is one of its strengths (and also a weakness)

**Comment by vrastov** [comment]
InteliliJ IDEA + google-style (https://google.github.io/styleguide/javaguide.html) https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml 

Without them, my example would look like this:

After first sorting: 
```xml
    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>foo.bar.veryLongArtifactId</artifactId>
    </dependency>

    <dependency>
      <groupId>foo.bar</groupId>
      <artifactId>baz</artifactId>
    </dependency>
```


```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.github.ekryd.sortpom.its</groupId>
  <artifactId>default-configuration</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>
        junit-jupiter-engine
      </artifactId>
      <version>5.7.0</version>
    </dependency>

  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.13.1</version>
        <executions>
          <execution>
            <goals>
              <goal>sort</goal>
            </goals>
            <configuration>
              <keepBlankLines>true</keepBlankLines>
              <createBackupFile>false</createBackupFile>
              <expandEmptyElements>false</expandEmptyElements>
              <sortModules>true</sortModules>
              <sortDependencies>scope,groupId,artifactId</sortDependencies>
              <sortProperties>true</sortProperties>
              <sortPlugins>groupId,artifactId</sortPlugins>
              <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```

executed `mvn clean verify`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.github.ekryd.sortpom.its</groupId>
  <artifactId>default-configuration</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.0</version>
    </dependency>

  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.13.1</version>
        <executions>
          <execution>
            <goals>
              <goal>sort</goal>
            </goals>
            <configuration>
              <keepBlankLines>true</keepBlankLines>
              <createBackupFile>false</createBackupFile>
              <expandEmptyElements>false</expandEmptyElements>
              <sortModules>true</sortModules>
              <sortDependencies>scope,groupId,artifactId</sortDependencies>
              <sortProperties>true</sortProperties>
              <sortPlugins>groupId,artifactId</sortPlugins>
              <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```
executed `mvn clean verify`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.github.ekryd.sortpom.its</groupId>
  <artifactId>default-configuration</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.0</version>
    </dependency>

  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.13.1</version>
        <executions>
          <execution>
            <goals>
              <goal>sort</goal>
            </goals>
            <configuration>
              <keepBlankLines>true</keepBlankLines>
              <createBackupFile>false</createBackupFile>
              <expandEmptyElements>false</expandEmptyElements>
              <sortModules>true</sortModules>
              <sortDependencies>scope,groupId,artifactId</sortDependencies>
              <sortProperties>true</sortProperties>
              <sortPlugins>groupId,artifactId</sortPlugins>
              <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>

```

**Comment by Ekryd** [comment]
So it is the google style plugin that reinserts the line breaks. No way around that unless you have shorter artifactIds.

But the other issue is a bug. Good job finding it!
The dependencies should be sorted correctly on the first pass, I'll get to it and try to solve it

**Comment by vrastov** [comment]
Added second example above

**Comment by Ekryd** [comment]
I have created a fix for the bug, please try it out.

* Clone the repo
* Switch to the correct branch: `git checkout issue-99-Invalid_sorting_on_line_break_in_artifactId_value`
* Build with `mvn clean install`
* Then you can use the sortpom plugin with the version `2.14.1-SNAPSHOT` in your project

Tell me how it works for you, so I can make an official release.

Btw. I have created integration tests for your two examples: 'test-linebreaks' and 'test-verify-linebreaks' under maven-plugin/src/it. Feel free to play with them :-D

**Comment by vrastov** [comment]
Use  mvn com.github.ekryd.sortpom:sortpom-maven-plugin:verify in CI/CD 

Until I see the difference, I'll see the manual launch.

**Comment by Ekryd** [comment]
Hi,
I understand that it is a bit of trouble to get the snapshot plugin into the CI-environment.
But excuse me for asking: where you able to test the snapshot version locally?

**Comment by vrastov** [comment]
Perhaps this is another mistake...

```cmd
rm -R ~/.m2/repository/com/github/ekryd/
git clone git@github.com:Ekryd/sortpom.git
...
cd sortpom/
git checkout issue-99-Invalid_sorting_on_line_break_in_artifactId_value
Ветка «issue-99-Invalid_sorting_on_line_break_in_artifactId_value» отслеживает внешнюю ветку «issue-99-Invalid_sorting_on_line_break_in_artifactId_value» из «origin».
Переключено на новую ветку «issue-99-Invalid_sorting_on_line_break_in_artifactId_value»
mvn clean install
....
rast@rast-msi:~/IdeaProjects/test-sortpom/sortpom$ ls  ~/.m2/repository/com/github/ekryd/*/*
/home/rast/.m2/repository/com/github/ekryd/echo-maven-plugin/echo-maven-plugin:
1.3.2

/home/rast/.m2/repository/com/github/ekryd/echo-maven-plugin/echo-output:
1.3.2

/home/rast/.m2/repository/com/github/ekryd/echo-maven-plugin/echo-plugin:
1.3.2

/home/rast/.m2/repository/com/github/ekryd/reflection-utils/reflection-utils:
1.1.0

/home/rast/.m2/repository/com/github/ekryd/sortpom/sortpom-maven-plugin:
2.14.1-SNAPSHOT  maven-metadata-local.xml

/home/rast/.m2/repository/com/github/ekryd/sortpom/sortpom-parent:
2.14.1-SNAPSHOT  maven-metadata-local.xml

/home/rast/.m2/repository/com/github/ekryd/sortpom/sortpom-sorter:
2.14.1-SNAPSHOT  maven-metadata-local.xml

git clone git@github.com:vrastov/test-sortpom.git
cd test-sortpom/
rast@rast-msi:~/IdeaProjects/test-sortpom/sortpom/test-sortpom$  mvn com.github.ekryd.sortpom:sortpom-maven-plugin:verify
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------< com.github.ekryd.sortpom.its:default-configuration >---------
[INFO] Building default-configuration 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- sortpom-maven-plugin:2.14.1-SNAPSHOT:verify (default-cli) @ default-configuration ---
[INFO] Verifying file /home/rast/IdeaProjects/test-sortpom/sortpom/test-sortpom/pom.xml
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.497 s
[INFO] Finished at: 2021-03-30T00:20:12+03:00
[INFO] ------------------------------------------------------------------------

rast@rast-msi:~/IdeaProjects/test-sortpom/sortpom/test-sortpom$ git status 
На ветке master
Ваша ветка обновлена в соответствии с «origin/master».

нечего коммитить, нет изменений в рабочем каталоге
ast@rast-msi:~/IdeaProjects/test-sortpom/sortpom/test-sortpom$ cat pom.xml 
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.github.ekryd.sortpom.its</groupId>
  <artifactId>default-configuration</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencies>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.0</version>
    </dependency>

    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>
        junit-jupiter-engine
      </artifactId>
      <version>5.7.0</version>
    </dependency>

  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>com.github.ekryd.sortpom</groupId>
        <artifactId>sortpom-maven-plugin</artifactId>
        <version>2.14.1-SNAPSHOT</version>
        <executions>
          <execution>
            <goals>
              <goal>sort</goal>
            </goals>
            <configuration>
              <keepBlankLines>true</keepBlankLines>
              <createBackupFile>false</createBackupFile>
              <expandEmptyElements>false</expandEmptyElements>
              <sortModules>true</sortModules>
              <sortDependencies>scope,groupId,artifactId</sortDependencies>
              <sortProperties>true</sortProperties>
              <sortPlugins>groupId,artifactId</sortPlugins>
              <spaceBeforeCloseEmptyElement>false</spaceBeforeCloseEmptyElement>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>

```

As I understand it, everything works.

Thanks!

**Comment by Ekryd** [comment]
Great!

And if you use the same test project with `mvn com.github.ekryd.sortpom:sortpom-maven-plugin:sort` then the only change is that
```xml
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>
        junit-jupiter-engine
      </artifactId>
      <version>5.7.0</version>
    </dependency>
```
becomes
```xml
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.0</version>
    </dependency>
```

Right?

**Comment by vrastov** [comment]
As I use the plugin, everything suits me.

In CI / CD - verify, sorting only by hand, followed by formatting by hand.

**Comment by vrastov** [comment]
I m closing issue?

**Comment by Ekryd** [comment]
Released SortPom version 2.14.1 with the bug fix. Enjoy!!
